File: src/csharp/examples/generic_reduce/Reduce.cs
using System;
using Alea.CUDA;
using Alea.CUDA.IL;
using Alea.CUDA.Utilities;

namespace Tutorial.Cs.examples.generic_reduce
{
    
    public class ReduceModule<T> : ILGPUModule
    {
        public static Func<int, T, T> MultiReduce(Func<T> initFunc, Func<T, T, T> reductionOp, int numWarps,
            int logNumWarps)
        {
            var warpStride = Const.WARP_SIZE + Const.WARP_SIZE/2 + 1;
            var sharedSize = numWarps*warpStride;

            return
                (tid, x) =>
                {
                    var warp = tid/Const.WARP_SIZE;
                    var lane = tid & (Const.WARP_SIZE - 1);
                    var shared =
                        Intrinsic.__ptr_volatile(
                            Intrinsic.__array_to_ptr(
                                __shared__.Array<T>(sharedSize)));
                    var warpShared = shared + warp*warpStride;
                    var s = warpShared + (lane + Const.WARP_SIZE/2);

                    warpShared[lane] = initFunc();
                    s[0] = x;

                    // Run inclusive scan on each warp's data.
                    var warpScan = x;
                    for (var i = 0; i < Const.LOG_WARP_SIZE; i++)
                    {
                        var offset = 1 << i;
                        warpScan = reductionOp.Invoke(warpScan, s[-offset]);
                        if (i < Const.LOG_WARP_SIZE - 1)
                            s[0] = warpScan;
                    }

                    var totalsShared =
                        Intrinsic.__ptr_volatile(
                            Intrinsic.__array_to_ptr(
                                __shared__.Array<T>(2*numWarps)));

                    // Last line of warp stores the warp scan.
                    if (lane == (Const.WARP_SIZE - 1))
                        totalsShared[numWarps + warp] = warpScan;

                    // Synchronize to make all the totals available to the reduction code.
                    Intrinsic.__syncthreads();

                    // Run an exclusive scan for the warp scans.
                    if (tid < numWarps)
                    {
                        // Grab the block total for the tid'th block.  This is the last element
                        // in the block's scanned sequence.  This operation avoids bank conflicts.
                        var total = totalsShared[numWarps + tid];
                        totalsShared[tid] = initFunc();
                        var ss = totalsShared + numWarps + tid;

                        var totalsScan = total;
                        for (var i = 0; i < logNumWarps; i++)
                        {
                            var offset = 1 << i;
                            totalsScan = reductionOp.Invoke(totalsScan, ss[-offset]);
                            ss[0] = totalsScan;
                        }
                    }

                    // Synchronize to make the block scan available to all warps.
                    Intrinsic.__syncthreads();

                    // The total is the last element.
                    return totalsShared[2*numWarps - 1];

                };
        }

        private readonly Func<T> _initFunc;
        private readonly Func<T,T,T> _reductionOp;
        private readonly Func<T,T> _transform;
        public Plan Plan;
        private readonly int _numThreads;
        private readonly Func<int, T, T> _multiReduce;

        public ReduceModule(GPUModuleTarget target, Func<T> initFunc, Func<T, T, T> reductionOp, Func<T, T> transform,
            Plan plan) : base(target)
        {
            _initFunc = initFunc;
            _reductionOp = reductionOp;
            _transform = transform;
            Plan = plan;
            _numThreads = plan.NumThreads;
            var numWarps = plan.NumWarps;
            var logNumWarps = Common.log2(numWarps);
            _multiReduce = MultiReduce(initFunc, reductionOp, numWarps, logNumWarps);
        }

        [Kernel]
        public void Upsweep(deviceptr<T> dValues, deviceptr<int> dRanges, deviceptr<T> dRangeTotals)
        {
            // Each block is processing a range.
            var range = blockIdx.x;
            var tid = threadIdx.x;
            var rangeX = dRanges[range];
            var rangeY = dRanges[range + 1];

            // Loop through all elements in the interval, adding up values.
            // There is no need to synchronize until we perform the multi-reduce.
            var reduced = _initFunc();
            var index = rangeX + tid;
            while (index < rangeY)
            {
                reduced = _reductionOp.Invoke(reduced, _transform.Invoke(dValues[index]));
                index += _numThreads;
            }

            // Get the total.
            var total = _multiReduce.Invoke(tid, reduced);

            if (tid == 0)
                dRangeTotals[range] = total;
        }

        [Kernel]
        public void ReduceRangeTotals(int numRanges, deviceptr<T> dRangeTotals)
        {
            var tid = threadIdx.x;
            var x = tid < numRanges ? dRangeTotals[tid] : _initFunc();
            var total = _multiReduce.Invoke(tid, x);

            if (tid == 0)
                dRangeTotals[0] = total;
        }

        public T Apply(T[] values)
        {
            var n = values.Length;
            var numSm = GPUWorker.Device.Attributes.MULTIPROCESSOR_COUNT;
            var tup = Plan.BlockRanges(numSm, n);
            var ranges = tup.Item1;
            var numRanges = tup.Item2;
            var lpUpsweep = new LaunchParam(numRanges, Plan.NumThreads);
            var lpReduce = new LaunchParam(1, Plan.NumThreadsReduction);

            using (var dValues = GPUWorker.Malloc(values))
            using (var dRanges = GPUWorker.Malloc(ranges))
            using (var dRangeTotals = GPUWorker.Malloc<T>(numRanges))
            {
                // Launch range reduction kernel to calculate the totals per range.
                GPUWorker.EvalAction(
                    () =>
                    {
                        GPULaunch(Upsweep, lpUpsweep, dValues.Ptr, dRanges.Ptr, dRangeTotals.Ptr);
                        if (numRanges > 1)
                        {
                            // Need to aggregate the block sums as well.
                            GPULaunch(ReduceRangeTotals, lpReduce, numRanges, dRangeTotals.Ptr);
                        }
                    });
                return dRangeTotals.Gather()[0];
            }

        }
    }
}

File: src/fsharp/examples/generic_reduce/Reduce.fs
(*** hide ***)
module Tutorial.Fs.examples.genericReduce.Reduce

open Microsoft.FSharp.Quotations
open Alea.CUDA
open Alea.CUDA.Utilities
open Plan

(**
Multi-reduce function for all warps in the block.
*)
(*** define:GenericReduceMultiReduce ***)
let multiReduce (initExpr:Expr<unit -> 'T>) (opExpr:Expr<'T -> 'T -> 'T>) numWarps logNumWarps =
    let warpStride = WARP_SIZE + WARP_SIZE / 2 + 1
    let sharedSize = numWarps * warpStride

    <@ fun tid (x:'T) ->
        let init = %initExpr
        let op = %opExpr
        let warp = tid / WARP_SIZE
        let lane = tid &&& (WARP_SIZE - 1)
        let shared = __shared__.Array<'T>(sharedSize) |> __array_to_ptr |> __ptr_volatile
        let warpShared = shared + warp * warpStride     
        let s = warpShared + (lane + WARP_SIZE / 2)

        warpShared.[lane] <- init()  
        s.[0] <- x

        // Run inclusive scan on each warp's data.
        let mutable warpScan = x
        for i = 0 to LOG_WARP_SIZE - 1 do
            let offset = 1 <<< i
            warpScan <- op warpScan s.[-offset]   
            if i < LOG_WARP_SIZE - 1 then s.[0] <- warpScan
        
        let totalsShared = __shared__.Array<'T>(2*numWarps) |> __array_to_ptr |> __ptr_volatile 

        // Last line of warp stores the warp scan.
        if lane = WARP_SIZE - 1 then totalsShared.[numWarps + warp] <- warpScan  

        // Synchronize to make all the totals available to the reduction code.
        __syncthreads()

        // Run an exclusive scan for the warp scans. 
        if tid < numWarps then
            // Grab the block total for the tid'th block. This is the last element
            // in the block's scanned sequence. This operation avoids bank conflicts.
            let total = totalsShared.[numWarps + tid]
            totalsShared.[tid] <- init()
            let s = totalsShared + numWarps + tid

            let mutable totalsScan = total
            for i = 0 to logNumWarps - 1 do
                let offset = 1 <<< i
                totalsScan <- op totalsScan s.[-offset]
                s.[0] <- totalsScan

        // Synchronize to make the block scan available to all warps.
        __syncthreads()

        // The total is the last element.
        totalsShared.[2 * numWarps - 1] @>

(**
Reduces ranges and store reduced values in array of the range totals.  
*)
(*** define:GenericReduceUpsweepKernel ***)   
let reduceUpSweepKernel (initExpr:Expr<unit -> 'T>) (opExpr:Expr<'T -> 'T -> 'T>) (transfExpr:Expr<'T -> 'T>) (plan:Plan) =
    let numThreads = plan.NumThreads
    let numWarps = plan.NumWarps
    let logNumWarps = log2 numWarps
    let multiReduce = multiReduce initExpr opExpr numWarps logNumWarps

    <@ fun (dValues:deviceptr<'T>) (dRanges:deviceptr<int>) (dRangeTotals:deviceptr<'T>) ->
        let init = %initExpr
        let op = %opExpr
        let transf = %transfExpr

        // Each block is processing a range.
        let range = blockIdx.x
        let tid = threadIdx.x
        let rangeX = dRanges.[range]
        let rangeY = dRanges.[range + 1]

        // Loop through all elements in the interval, adding up values.
        // There is no need to synchronize until we perform the multireduce.
        let mutable reduced = init()
        let mutable index = rangeX + tid
        while index < rangeY do              
            reduced <- op reduced (transf dValues.[index]) 
            index <- index + numThreads

        // Get the total.
        let total = (%multiReduce) tid reduced 

        if tid = 0 then dRangeTotals.[range] <- total @>

(**
Reduces range totals to a single total, which is written back to the first element in the range totals input array.
*)
(*** define:GenericReduceRangeTotalsKernel ***)
let reduceRangeTotalsKernel (initExpr:Expr<unit -> 'T>) (opExpr:Expr<'T -> 'T -> 'T>) (plan:Plan) =
    let numThreads = plan.NumThreadsReduction
    let numWarps = plan.NumWarpsReduction
    let logNumWarps = log2 numWarps
    let multiReduce = multiReduce initExpr opExpr numWarps logNumWarps

    <@ fun numRanges (dRangeTotals:deviceptr<'T>) ->
        let init = %initExpr
        let op = %opExpr

        let tid = threadIdx.x
        let x = if tid < numRanges then dRangeTotals.[tid] else (init())
        let total = (%multiReduce) tid x

        // Have the first thread in the block set the total and store it in the first element of the input array.
        if tid = 0 then dRangeTotals.[0] <- total @>

